README.md
# Student CLI
![GitHub](https://img.shields.io/github/license/sci-oer/student-cli?style=plastic)
![GitHub](https://img.shields.io/github/license/sci-oer/tudent-cli?style=plastic)


Wrapper script to help start and stop the sci-oer docker container.


## System Requirements
[System Requirements]: #system-requirements

- python 3.7 or later installed
- Docker must be installed and running

## Installation

The following sections cover the various methods to set up the sci-oer student cli.

### Installing from PyPI or GitHub
[Installing from PyPI or GitHub]: #installing-from-pypi-or-github

[pip] can be used to install scioer from the Python Package Index (PyPI).
It will also install the necessary Python libraries.


```bash
# the latest stable version
python3 -m pip3 install sci-oer

# a specific version (e.g. 1.0.0)
python3 -m pip3 install sci-oer==1.0.0
```

If [git] is installed, pip can also install the latest snapshot of the official
scioer source code repository on GitHub.

```bash
# the latest snapshot of the official source code repository (requires git)
python3 -m pip3 install git+https://github.com/sci-oer/student-cli#egg=scioer
```

[git]: https://git-scm.com/
[pip]: https://pip.pypa.io/en/stable/


### Working Locally

When working on the project itself, it is sometimes useful to set up a local development environment, where it is possible to directly run the CLI and the test suite, without building and installing a local package.

For this, start by installing [git] and any system-level dependencies mentioned in [System Requirements]. Then, clone the repository and change into the created directory:

```bash
git clone https://github.com/sci-oer/student-cli
cd student-cli
```

Optionally, set up a [virtual environment].

Finally, if the necessary Python build, test and runtime libraries are not already installed on the environment (virtual or global), manually install them:

```bash
pip3 install .
```

At this point, the environment is set up. To run the test suite, execute:

```bash
python3 -m pytest
```
To run the CLI directly, without building and installing a local package, execute:

```bash
python3 -m scioer [arguments]
```
And to install scioer into the environment:

```bash
pip3 install .
```


#### Creating a virtual environment
[Creating a virtual environment]: #creating-a-virtual-environment

Setting up a virtual environment is an optional step.  Even so, installing
Python packages directly in the global environment is not generally advised.

Instead, it is usual to first set up a [virtual environment]:

```bash
# create virtual enviroment at <path>
python -m venv <path>
```

Once set up, the virtual environment can be activated on the current shell
(more information in the [official documentation][virtual environment]).
Alternatively, the virtual environment can also be used directly, without
activation, by prefixing all `python` and `pip` invocations with the
environment's bin directory.

```bash
# Linux/macOS/BSDs (POSIX)
<path>/bin/python [arguments]
<path>/bin/pip [arguments]

# Windows
<path>\Scripts\python [arguments]
<path>\Scripts\pip [arguments]
```

[virtual environment]: https://docs.python.org/3/library/venv.html

## The Command Line Interface

The complete list of commands and options can be found in `sci-oer --help` and in the man page, but the following topics cover the most common operations.

Brackets `[ ]`, parenthesis `( )`, less than/greater than `< >` and ellipsis `...` are used to describe, respectively, optional, required, positional and repeating elements.
Example commands are prefixed with a number sign `#`, which also serves to indicate that on Linux root permissions (or suitable group membership) may be required.

The `--verbose` option will print some extra information, like automatically made adjustments to user-provided settings.
And if there is a problem, the `--debug` flag will make `sci-oer` output more information to help identify its cause; be sure to include this when opening a new issue.

## Additional Documentation

** TODO **

## Related projects
[Related projects]: #related-projects

### [sci-oer/base-resource](https://github.com/sci-oer/base-resource)

The base open educational resource Docker image.
- [sci-oer/java-resource](https://github.com/sci-oer/java-resource)
- [sci-oer/c-resource](https://github.com/sci-oer/c-resource)
- [sci-oer/mariadb-resource](https://github.com/sci-oer/mariadb-resource)


### [sci-oer/automated-builder](https://github.com/sci-oer/automated-builder)

The [cli tool] and [Docker image] to create customized versions of the base resource images.


[cli tool]: https://pypi.org/project/scioer-builder/
[Docker image]: https://hub.docker.com/r/scioer/automated-builder

## Contributing

Contributions are more than welcome on this project!
Check out the list of [issues](https://github.com/sci-oer/student-cli) for tasks that you want to try implementing.

Check out the [CONTRIBUTING.md](.github/CONTRIBUTING.md) guide for more information about how to contribute to the project.


## Issues Getting Support

Like any software tool things go wrong and sometimes there are issues.
This section presents some common issues that can arise and solutions to resolve the issue.
If these topics are not able to help you resolve the issue feel free to open a [support request], and we be happy to help.

[support request]: https://github.com/sci-oer/student-cli/issues/new

### I did some work in the container, and now I can't find my files

If Any files that you edited in the `/course` directory within the container will be saved to your computer in one of the following locations:

1. `~/Desktop/<container name>`
2. `$XDG_DATA_HOME/scioer/<container name>`
3. `$HOME/.local/share/scioer/<container name>`

Or it will be saved to the location that was specified through the config file or the cli flags that were specified when the container was started.

If your file was not created under the `/course` directory then make sure that you move it into that folder _before_ the container is stopped and removed, otherwise it will be permanently lost.

Do not work on any of your files in the `~` directory from within the container those files will be lost when the container is recreated.

### Permission denied error when trying to edit the files

If you are trying to access the files from your computer that you created or edited from within the container and are now getting permission denied errors.
When this happens it is likely because the container is running as a different user.
This can be checked by running the following commands:

```bash
$ id -u    # This will print the user id of the current user
501

$ id -g    # This will print the group id of the current user
20

$ ls -nd <path to mount >   # The 3rd and 4th columns will be the user id and group id that own the files
drwxr-xr-x  14 501  20  448  3 Apr 13:33 .
                 ^   ^
                 |   |
                 | group Id
                User Id
```

If these do not match you can `sudo chown -R <userId>:<groupId> <volume dir>` to change the ownership of the files.
Then you will need to update the configuration of `sci-oer` to launch the container as the correct user.

### I started the container and all of my work seems to be gone

Make sure that you save all of your work in the `/course` directory, so it can be saved to the persistent volume on your computer.
If you did not save it to the correct spot and the container has not yet been removed then you can still enter the container and move the files, so they can be saved.

### I went to `localhost:3000`, and it says there is no page

- Did you double-check that the container is running? `docker ps`
- Is `3000` the correct port? `docker ps` or `docker inspect <container name>`



CHANGELOG.md
# Changelog

## [v1.3.1](https://github.com/sci-oer/student-cli/releases/v1.3.1) (2023-09-05)

[Full Changelog](https://github.com/sci-oer/student-cli/compare/v1.3.0...v1.3.1)

**Merged pull requests:**

- chore\(deps\): update actions/checkout action to v4 [\#172](https://github.com/sci-oer/student-cli/pull/172) ([renovate[bot]](https://github.com/apps/renovate))
- chore\(deps\): update dependency shellingham to v1.5.3 [\#171](https://github.com/sci-oer/student-cli/pull/171) ([renovate[bot]](https://github.com/apps/renovate))
- chore\(deps\): update dependency pytest to v7.4.1 [\#170](https://github.com/sci-oer/student-cli/pull/170) ([renovate[bot]](https://github.com/apps/renovate))
- chore\(deps\): update dependency click to v8.1.7 [\#169](https://github.com/sci-oer/student-cli/pull/169) ([renovate[bot]](https://github.com/apps/renovate))
- chore\(deps\): update pypa/gh-action-pypi-publish action to v1.8.10 [\#168](https://github.com/sci-oer/student-cli/pull/168) ([renovate[bot]](https://github.com/apps/renovate))
- chore\(deps\): update dependency pyparsing to v3.1.1 [\#167](https://github.com/sci-oer/student-cli/pull/167) ([renovate[bot]](https://github.com/apps/renovate))
- chore\(deps\): update dependency rich to v13.5.2 [\#166](https://github.com/sci-oer/student-cli/pull/166) ([renovate[bot]](https://github.com/apps/renovate))
- chore\(deps\): update dependency pathspec to v0.11.2 [\#165](https://github.com/sci-oer/student-cli/pull/165) ([renovate[bot]](https://github.com/apps/renovate))
- chore\(deps\): update dependency wheel to v0.41.2 [\#163](https://github.com/sci-oer/student-cli/pull/163) ([renovate[bot]](https://github.com/apps/renovate))
- chore\(deps\): update dependency platformdirs to v3.10.0 [\#153](https://github.com/sci-oer/student-cli/pull/153) ([renovate[bot]](https://github.com/apps/renovate))

## [v1.3.0](https://github.com/sci-oer/student-cli/releases/v1.3.0) (2023-07-27)

[Full Changelog](https://github.com/sci-oer/student-cli/compare/v1.2.0...v1.3.0)

**Merged pull requests:**

- chore\(deps\): update dependency pyyaml to v6.0.1 [\#160](https://github.com/sci-oer/student-cli/pull/160) ([renovate[bot]](https://github.com/apps/renovate))
- chore\(deps\): update dependency click to v8.1.6 [\#159](https://github.com/sci-oer/student-cli/pull/159) ([renovate[bot]](https://github.com/apps/renovate))
- chore\(deps\): update dependency pyparsing to v3.1.0 [\#158](https://github.com/sci-oer/student-cli/pull/158) ([renovate[bot]](https://github.com/apps/renovate))
- chore\(deps\): update dependency pytest to v7.4.0 [\#157](https://github.com/sci-oer/student-cli/pull/157) ([renovate[bot]](https://github.com/apps/renovate))
- chore\(deps\): update dependency rich to v13.4.2 [\#156](https://github.com/sci-oer/student-cli/pull/156) ([renovate[bot]](https://github.com/apps/renovate))
- chore\(deps\): update dependency typing\_extensions to v4.7.1 [\#155](https://github.com/sci-oer/student-cli/pull/155) ([renovate[bot]](https://github.com/apps/renovate))
- chore\(deps\): update dependency setuptools to v67.8.0 [\#154](https://github.com/sci-oer/student-cli/pull/154) ([renovate[bot]](https://github.com/apps/renovate))
- chore\(deps\): update dependency docker to v6.1.3 [\#151](https://github.com/sci-oer/student-cli/pull/151) ([renovate[bot]](https://github.com/apps/renovate))
- chore\(deps\): update pypa/gh-action-pypi-publish action to v1.8.8 [\#150](https://github.com/sci-oer/student-cli/pull/150) ([renovate[bot]](https://github.com/apps/renovate))
- chore\(deps\): update dependency typer to v0.9.0 [\#149](https://github.com/sci-oer/student-cli/pull/149) ([renovate[bot]](https://github.com/apps/renovate))

## [v1.2.0](https://github.com/sci-oer/student-cli/releases/v1.2.0) (2023-04-30)

[Full Changelog](https://github.com/sci-oer/student-cli/compare/v1.1.0...v1.2.0)

**Implemented enhancements:**

- feat: add readline so prompts can be edited [\#148](https://github.com/sci-oer/student-cli/pull/148) ([MarshallAsch](https://github.com/MarshallAsch))

**Merged pull requests:**

- chore!: turn off windows ci piplines that keep failing [\#147](https://github.com/sci-oer/student-cli/pull/147) ([MarshallAsch](https://github.com/MarshallAsch))
- chore: update release notes template [\#146](https://github.com/sci-oer/student-cli/pull/146) ([MarshallAsch](https://github.com/MarshallAsch))
- chore: auto tag renovate prs as dependancy updates [\#145](https://github.com/sci-oer/student-cli/pull/145) ([MarshallAsch](https://github.com/MarshallAsch))
- chore\(deps\): update dependency attrs to v23 [\#144](https://github.com/sci-oer/student-cli/pull/144) ([renovate[bot]](https://github.com/apps/renovate))
- chore\(deps\): update dependency packaging to v23.1 [\#143](https://github.com/sci-oer/student-cli/pull/143) ([renovate[bot]](https://github.com/apps/renovate))
- chore\(deps\): update dependency black to v23.3.0 [\#142](https://github.com/sci-oer/student-cli/pull/142) ([renovate[bot]](https://github.com/apps/renovate))
- chore\(deps\): update dependency pathspec to v0.11.1 [\#141](https://github.com/sci-oer/student-cli/pull/141) ([renovate[bot]](https://github.com/apps/renovate))
- chore\(deps\): update dependency wheel to v0.40.0 [\#140](https://github.com/sci-oer/student-cli/pull/140) ([renovate[bot]](https://github.com/apps/renovate))
- chore\(deps\): update pypa/gh-action-pypi-publish action to v1.8.5 [\#139](https://github.com/sci-oer/student-cli/pull/139) ([renovate[bot]](https://github.com/apps/renovate))
- chore\(deps\): update dependency setuptools to v67.7.2 [\#138](https://github.com/sci-oer/student-cli/pull/138) ([renovate[bot]](https://github.com/apps/renovate))
- chore\(deps\): update dependency rich to v13.3.5 [\#137](https://github.com/sci-oer/student-cli/pull/137) ([renovate[bot]](https://github.com/apps/renovate))

tests/test_cli.py
from typer.testing import CliRunner

from scioer.cli import app

runner = CliRunner()


def test_app():
    result = runner.invoke(app, ["--help"])
    assert result.exit_code == 0
    # assert "Hello name" in result.stdout


# assert "Let's have a coffee in Berlin" in result.stdout


scioer/docker.py
from typing import Container
import docker
import logging
import os
import typer
import subprocess
import re
from collections.abc import Mapping
import sys

_LOGGER = logging.getLogger(__name__)


def port_mapping(mapping: str, public: bool) -> Mapping:
    m = re.fullmatch("^(([0-9]{1,5})(?:/(?:tcp|udp))?):([0-9]{1,5})$", mapping)
    if not m:
        typer.secho(
            f"Invalid port specification '{mapping}'",
            fg=typer.colors.RED,
        )
        raise typer.Exit(code=1)

    container = m.group(1)
    srcPort = int(m.group(2))
    hostPort = int(m.group(3))

    if srcPort < 0 or srcPort > 65535:
        typer.secho(
            f"Invalid port number '{srcPort}'",
            fg=typer.colors.RED,
        )
        raise typer.Exit(code=1)

    if hostPort < 0 or hostPort > 65535:
        typer.secho(
            f"Invalid port number '{srcPort}'",
            fg=typer.colors.RED,
        )
        raise typer.Exit(code=1)

    host = "0.0.0.0" if public else "127.0.0.1"
    return {container: (host, hostPort) if hostPort != 0 else None}


def port_map(portList: list, public: bool) -> dict:
    portMapping = {}
    for p in portList:
        portMapping.update(port_mapping(p, public))

    return portMapping


def port_env_mapping(mapping: str) -> str:
    m = re.fullmatch("^(([0-9]{1,5})(?:/(?:tcp|udp))?):([0-9]{1,5})$", mapping)
    if not m:
        typer.secho(
            f"Invalid port specification '{mapping}'",
            fg=typer.colors.RED,
        )
        raise typer.Exit(code=1)

    srcPort = int(m.group(2))
    hostPort = int(m.group(3))

    if srcPort < 0 or srcPort > 65535:
        typer.secho(
            f"Invalid port number '{srcPort}'",
            fg=typer.colors.RED,
        )
        raise typer.Exit(code=1)

    if hostPort < 0 or hostPort > 65535:
        typer.secho(
            f"Invalid port number '{srcPort}'",
            fg=typer.colors.RED,
        )
        raise typer.Exit(code=1)

    return f"PORT_{srcPort}={hostPort}"


def port_env_map(portList: list) -> list:
    return [port_env_mapping(p) for p in portList]


def fetch_latest(client: docker.client, repository, **kwargs):
    _LOGGER.info(
        f'pulling latest version of the "{repository}" docker image, this may take a while...'
    )

    oldImage = None
    try:
        oldImage = client.images.get(repository)
    except:
        pass
    image = client.images.pull(repository)
    _LOGGER.info("Done pulling the latest docker image")

    if oldImage and oldImage.id != image.id:
        oldImage.remove()
    return image


def create_container(client: docker.client, course: dict, **kwargs):
    try:
        _LOGGER.info(f"checking if image {course['image']} exists locally...")
        i = client.images.get(course["image"])
        _LOGGER.info("Image exists locally.")
    except docker.errors.ImageNotFound as e:
        _LOGGER.info("Image is not found, start will take a while to pull first.")
        typer.secho(
            f"Course image needs to be downloaded, this may take a while...",
            fg=typer.colors.YELLOW,
        )

    _LOGGER.info(f"starting `{course['image']}` container as `{course['name']}`...")
    try:
        container = client.containers.run(
            course["image"],
            ports=port_map(course["ports"], course.get("public", False)),
            environment=port_env_map(course["ports"]),
            name=f'scioer_{course["name"]}',
            hostname=course["name"],
            tty=True,
            detach=True,
            volumes=[f"{course['volume']}:/course"],
        )
    except docker.errors.ImageNotFound as e:
        _LOGGER.error("Image not found.", e)
        typer.secho(
            f"Course image not found, check the config file that the image name is correct.",
            fg=typer.colors.RED,
        )
        sys.exit(1)
    except docker.errors.APIError as e:
        _LOGGER.debug(f"Failed to start the container: {e}")

        if e.status_code == 409:
            typer.secho(
                f"Container name already in use. Please delete the container with the name `scioer_{course['name']}` before trying again.",
                fg=typer.colors.RED,
            )
            sys.exit(2)
        elif e.status_code == 404:
            typer.secho(
                f"Course image not found, check the config file that the image name is correct.",
                fg=typer.colors.RED,
            )
            sys.exit(3)

        typer.secho(
            f"Unknown error: {e.explanation}, aborting...",
            fg=typer.colors.RED,
        )
        sys.exit(4)

    return container


def start_container(client: docker.client, course: dict, **kwargs):
    container = None
    try:
        container = client.containers.get(f'scioer_{course["name"]}')
        _LOGGER.info(f'Container for `scioer_{course["name"]}` already exists.')

        if container.status == "running":
            _LOGGER.info("Container is already running, not restarting.")
        else:
            _LOGGER.info("Restarting container")
            container.start()
            _LOGGER.info("Successfully started")

    except:
        _LOGGER.info(f'Container `scioer_{course["name"]}` does not exist, starting...')
        container = create_container(client, course)

    return container


def stop_container(client: docker.client, courseName: str, keep: bool, **kwargs):
    _LOGGER.info("stopping docker container...")

    try:
        container = client.containers.get(f"scioer_{courseName}")
    except:
        typer.secho(
            f"Container for course '{courseName}' is not running",
            fg=typer.colors.YELLOW,
        )
        return

    container.stop()
    typer.secho(
        f"Container for course '{courseName}' is has been stopped",
        fg=typer.colors.GREEN,
    )

    if not keep:
        delete_container(container)


def attach(client: docker.client, courseName: str, **kwargs):
    _LOGGER.info("attaching to docker container...")

    try:
        container = client.containers.get(f"scioer_{courseName}")
    except:
        typer.secho(
            f"Container for course '{courseName}' is not running",
            fg=typer.colors.YELLOW,
        )
        return

    os.system(f"docker exec -it scioer_{courseName} cat /scripts/motd.txt")

    typer.echo("Starting interactive shell in the container, type `exit` to quit.")
    os.system(f"docker exec -it scioer_{courseName} bash --login")


def delete_container(container, **kwargs):
    _LOGGER.info("Deleting container...")
    container.remove()


def setup():
    client = None
    try:
        client = docker.from_env()
    except:
        typer.secho(
            "failed to connect to docker, check that Docker is running on the host.",
            fg=typer.colors.RED,
        )
        raise typer.Exit(code=1)

    return client


scioer/cli.py
import typer
from collections.abc import Mapping


import click
import scioer.config.load as load
import scioer.config.parse as parser
import os
import re
from typing import Optional
from pathlib import Path
import logging

try:
    import readline
except:
    import sys

    if sys.platform == "win32" or sys.platform == "cygwin":
        try:
            from pyreadline3 import Readline
        except:
            pass

__version__ = "UNKNOWN"
try:
    from scioer.__version__ import __version__
except:
    pass

import scioer.docker as docker

_LOGGER = logging.getLogger(__name__)

app = typer.Typer(
    name="Self Contained Interactive Open Educational Resource Helper",
    help=""" A CLI tool to help configure, start, stop course resources.

    \b
    Common usage commands:
    1. `scioer config`
        .... fill in the form
    2. `scioer start <course>`
    3. `scioer shell <course>`
    4. `scioer stop <course>`
    \f
    """,
    no_args_is_help=True,
    context_settings={"help_option_names": ["-h", "--help"]},
)


def conf_callback(ctx: typer.Context, param: typer.CallbackParam, value: Path):
    if value:
        value = os.path.realpath(os.path.expanduser(str(value)))

    configFiles = load.get_config_files(value)
    config = load.filter_config_files(configFiles)

    if not value and not config:
        config = configFiles[0]
        _LOGGER.info(f"No config file found, using default: {config}")
    elif value and value != config:
        config = value
        _LOGGER.info(f"Config file does not exist yet, using anyway: {config}")

    if config:
        _LOGGER.info(f"Loading from config file: {config}")

        data = parser.load_config_file(config)

        ctx.default_map = ctx.default_map or {}  # Initialize the default map
        ctx.default_map.update(data)  # Merge the config dict into default_map

        ctx.obj = ctx.obj or {}  # Initialize the object
        ctx.obj.update(
            {"config_file": config, "config": data}
        )  # Merge the config dict into object

    return config


def version_callback(value: bool):
    if value:
        typer.echo(f"scioer CLI Version: {__version__}")
        raise typer.Exit()


configOption = typer.Option(
    None,
    "--config",
    "-c",
    metavar="FILE",
    dir_okay=False,
    resolve_path=False,
    readable=True,
    writable=True,
    callback=conf_callback,
    is_eager=True,
    help="Path to the yaml config file",
)

courseNameArgument = typer.Argument(
    None, metavar="COURSE_NAME", help="The name of the course"
)


def load_course(config: Mapping, courseName: str, ask: bool = True):
    course = config.get(courseName, {})

    while ask and not course:
        if courseName:
            typer.secho(
                f'Course "{courseName} is not found. use `scioer config` if you want to create it.',
                fg=typer.colors.YELLOW,
            )
        courses = [k for k in config.keys() if isinstance(config[k], dict)]
        courseName = click.prompt(
            "Course not found, did you mean one of:", type=click.Choice(courses)
        )

        course = config.get(courseName, {})

    if course:
        course["name"] = courseName

    _LOGGER.info(f"course content: {course}")
    return course


def print_post_start_help(courseName):
    typer.secho(f"Started the {courseName} course resource", fg=typer.colors.GREEN)
    typer.secho(
        f"Login for the wiki: 'admin@example.com' 'password'", fg=typer.colors.GREEN
    )

    typer.echo("-----")
    typer.secho(f"To stop the course: scioer stop {courseName}", fg=typer.colors.YELLOW)
    typer.secho(
        f"To get a shell for the course: scioer shell {courseName}",
        fg=typer.colors.YELLOW,
    )
    typer.secho(
        f"To re-start the course: scioer start {courseName}", fg=typer.colors.YELLOW
    )
    typer.secho(
        f"To get information on the courses: scioer status", fg=typer.colors.YELLOW
    )


@app.command()
def start(
    ctx: typer.Context,
    name: Optional[str] = courseNameArgument,
    pull: bool = False,
    configFile: Optional[Path] = configOption,
):
    """Start a oer container"""

    client = docker.setup()

    config = ctx.default_map
    if not name and len(config.keys()) == 1:
        name = list(config.keys())[0]

    course = load_course(config, name)

    if course.get("auto_pull", False) or pull:
        typer.secho(
            f"Pulling the latest version of the {course['name']}",
            fg=typer.colors.GREEN,
        )
        typer.secho(
            f"This may take a while...",
            fg=typer.colors.YELLOW,
        )
        docker.fetch_latest(client, course["image"])

    typer.secho("Starting...", fg=typer.colors.GREEN)
    docker.start_container(client, course)
    print_post_start_help(course["name"])


@app.command()
def stop(
    ctx: typer.Context,
    name: Optional[str] = courseNameArgument,
    keep: Optional[bool] = typer.Option(False, "--no-remove", "-k"),
    configFile: Optional[Path] = configOption,
):
    """Stop a running course container"""

    client = docker.setup()

    config = ctx.default_map
    if not name and len(config.keys()) == 1:
        name = list(config.keys())[0]

    course = load_course(config, name, ask=False)

    if course:
        typer.secho(
            f"Stopping course container, this make take a couple seconds...",
            fg=typer.colors.RED,
        )
        docker.stop_container(client, course["name"], keep)
    else:
        typer.secho(f'Course "{name}" is not running.', fg=typer.colors.YELLOW)


@app.command()
def shell(
    ctx: typer.Context,
    name: str = courseNameArgument,
    configFile: Optional[Path] = configOption,
):
    """Start a shell in a running course container"""

    client = docker.setup()

    config = ctx.default_map
    if not name and len(config.keys()) == 1:
        name = list(config.keys())[0]

    course = load_course(config, name, ask=False)

    if course:
        docker.attach(client, course["name"])
    else:
        typer.secho(f'Course "{name}" is not running.', fg=typer.colors.YELLOW)


def print_container(container):
    indent = 2
    print(f"Course: {container.name[7:]} ({container.status})")

    print(f'{" " * indent }Ports:')
    ports = list(container.ports.items())
    if ports:
        for port in ports[:-1]:
            print(f'{" " * indent }├── {port[1][0]["HostPort"]} -> {port[0][:-4]}')
        port = ports[-1]
        print(f'{" " * indent }└── {port[1][0]["HostPort"]} -> {port[0][:-4]}')

    print(f'{" " * indent }Volumes:')
    volumes = [v for v in container.attrs["Mounts"] if v["Type"] == "bind"]

    home = os.path.expanduser("~")
    if volumes:
        for volume in volumes[:-1]:
            hostPath = volume["Source"].replace(home, "~")
            print(f'{" " * indent }├── {hostPath} as {volume["Destination"]}')

        volume = volumes[-1]
        hostPath = volume["Source"].replace(home, "~")
        print(f'{" " * indent }└── {hostPath} as {volume["Destination"]}')


@app.command()
def status(
    ctx: typer.Context,
    configFile: Optional[Path] = configOption,
):
    """Prints all the information about the running container"""
    client = docker.setup()

    config = ctx.default_map
    config_file = ctx.obj["config_file"]
    _LOGGER.info(f"Config contents: {config}")
    names = list(config.keys())

    home = os.path.expanduser("~")
    configPath = config_file.replace(home, "~")
    print(f"Config file: {configPath}")

    # containers = client.containers.list(all=True)
    # containers = [c for c in containers if c.name.replace('scioer_', '') in names ]

    filtered = []
    notRunning = []
    for n in names:
        try:
            c = client.containers.get(f"scioer_{n}")
            _LOGGER.debug(f"Container information for course {n}: {c.attrs}")
            filtered.append(c)
        except:
            notRunning.append(n)

    for c in filtered:
        print_container(c)

    for n in notRunning:
        print(f"Course: {n} (Not Running)")


def prompt_port(message: str, default: int) -> int:
    value = typer.prompt(
        message,
        default=default,
        type=int,
    )

    while value < 0 or value > 65535:
        typer.secho(
            f"`{value}` is not a valid port number.",
            fg=typer.colors.RED,
        )

        value = typer.prompt(
            message,
            default=default,
            type=int,
        )
    return value


def port_mapping(mapping: str) -> Mapping:
    m = re.fullmatch("^(([0-9]{1,5})(?:/(?:tcp|udp))?)(?::([0-9]{1,5}))?$", mapping)
    if not m:
        return {}

    container = m.group(1)
    srcPort = int(m.group(2))
    hostPort = int(m.group(3) or m.group(2))

    if srcPort < 0 or srcPort > 65535 or hostPort < 0 or hostPort > 65535:
        typer.secho(
            "Invalid port number.",
            fg=typer.colors.RED,
        )
        return {}

    return {container: hostPort}


def prompt_custom_ports() -> Mapping:
    value = typer.prompt(
        "Custom ports to expose, in the form of 'container[:host]', or no input to skip ",
        default="",
        value_proc=lambda v: v.strip(),
        type=str,
    )

    mapping = port_mapping(value)
    if value != "" and not mapping:
        typer.secho(
            "Invalid port specification, please try again.",
            fg=typer.colors.RED,
        )

    mappings = mapping

    while value != "":
        value = typer.prompt(
            "Custom ports to expose, in the form of 'container:host', or no input to skip ",
            default="",
            value_proc=lambda v: v.strip(),
            type=str,
        )

        mapping = port_mapping(value)
        if value != "" and not mapping:
            typer.secho(
                "Invalid port specification, please try again.",
                fg=typer.colors.RED,
            )
            continue

        if mapping:
            mappings = {**mappings, **mapping}

    return mappings


@app.command()
def config(
    ctx: typer.Context,
    configFile: Optional[Path] = configOption,
):
    """Setup a new course resource, or edit an existing one"""

    if not configFile:
        typer.echo(
            f"Config file not found, or not specified. Make sure that file exists or use `--config=FILE` to specify the file"
        )
        raise typer.Exit(1)

    config = ctx.default_map
    _LOGGER.info(f"config contents: {config}")

    course_name = typer.prompt("What's the name of the course?")
    safe_course_name = "".join(
        x for x in course_name.replace(" ", "_") if x.isalnum() or x == "_"
    )

    default_image = "scioer/oo-java:W23"
    default_volume = os.path.join(os.path.expanduser("~/Desktop"), safe_course_name)

    course = config.get(safe_course_name, {})

    docker_image = typer.prompt(
        "What docker image does the course use?",
        default=course.get("image", default_image),
    )

    auto_pull = typer.confirm(
        "Automatically fetch new versions", default=course.get("auto_pull", False)
    )

    course_storage = typer.prompt(
        "Where should the files for the course be stored?",
        default=course.get("volume", default_volume),
    )

    useDefaults = typer.confirm("Use the default ports", default=True)

    mappings = {
        "3000": 3000,
        "8888": 8888,
        "8000": 8000,
        "22": 2222,
    }

    if not useDefaults:
        wiki_port = prompt_port(
            "Wiki port to expose, (0 to publish a random port)",
            3000,
        )

        jupyter_port = prompt_port(
            "Jupyter notebooks port to expose, (0 to publish a random port)",
            8888,
        )

        lectures_port = prompt_port(
            "Lectures port to expose, (0 to publish a random port)",
            8000,
        )

        ssh_port = prompt_port(
            "ssh port to expose, (0 to publish a random port)",
            2222,
        )

        customPorts = prompt_custom_ports()

        mappings = {
            "3000": wiki_port,
            "8888": jupyter_port,
            "8000": lectures_port,
            "22": ssh_port,
            **customPorts,
        }

    ports = [f"{k}:{v}" for k, v in mappings.items()]

    config[safe_course_name] = {
        "image": docker_image,
        "volume": os.path.realpath(os.path.expanduser(course_storage)),
        "ports": ports,
        "auto_pull": auto_pull,
        "public": False,
    }

    parser.save_config_file(configFile, config)


@app.callback()
def setup(
    verbose: Optional[bool] = typer.Option(False, "--verbose", "-v"),
    debug: Optional[bool] = typer.Option(False, "--debug"),
    version: Optional[bool] = typer.Option(
        None, "--version", "-V", callback=version_callback, is_eager=True
    ),
):
    level = logging.WARNING
    if debug:
        level = logging.DEBUG
    elif verbose:
        level = logging.INFO

    logging.basicConfig(level=level)


if __name__ == "__main__":
    app()
